### PR TITLE
Need "Find all References" in Form Designer context menu

### LINKED ISSUE #3260: Need "Find all References" in Form Designer context menu
Currently the only RD command in the form designer is "Rename". It would be nice to have "Find all References" as well, so in-code references to a given form control can be easily located straight from the designer view.

### PR DESCRIPTION
**Pull request by mansellan**
Closes #3260

### CODE REVIEW

File: RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs [hunk 1]
```diff
@@ -87,7 +88,8 @@ private void UpdateTab()
 
         protected override bool EvaluateCanExecute(object parameter)
         {
-            if (_vbe.ActiveCodePane == null || _state.Status != ParserState.Ready)
+            if ((_state?.Status ?? ParserState.None) != ParserState.Ready ||
+                (_vbe.ActiveCodePane == null && !(_vbe.SelectedVBComponent?.HasDesigner ?? false)))
```
Comment (retailcoder) [inline RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs]: If `_state` is `null`, we have one hell of a problem - it's safe to assume it's not null. That null-checking seems overkill.

File: RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs [hunk 2]
```diff
@@ -100,7 +102,7 @@ protected override bool EvaluateCanExecute(object parameter)
 
         protected override void OnExecute(object parameter)
         {
-            if (_state.Status != ParserState.Ready)
+            if ((_state?.Status ?? ParserState.None) != ParserState.Ready)
```
Comment (retailcoder) [inline RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs]: I'm missing the *why* here - have you encountered a case where `_state` was `null`? I'm curious...

File: RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs (hunk 1 above)
Comment (mansellan) [reply RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs]: Fair enough, will remove. I added based on nullchecks I saw elsewhere.

Comment (retailcoder) [reply RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs]: Let me know where they are if you encounter them - they should probably go as well :-)

File: RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs [hunk 3]
```diff
@@ -154,15 +156,58 @@ private SearchResultsViewModel CreateViewModel(Declaration declaration)
...
+        {
             return _state.FindSelectedDeclaration(_vbe.ActiveCodePane);
         }
 
+        private Declaration FindFormDesignerTarget(QualifiedModuleName? qualifiedModuleName = null)
+        {            
+            (var projectId, var component)
+                = qualifiedModuleName.HasValue
+                                     ? (qualifiedModuleName.Value.ProjectId, qualifiedModuleName.Value.Component)
+                                     : (_vbe.ActiveVBProject.ProjectId, _vbe.SelectedVBComponent);
```
Comment (retailcoder) [inline RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs]: That looks a bit weird. How about this:

```csharp
(var projectId, var component) = qualifiedModuleName.HasValue
    ? (qualifiedModuleName.Value.ProjectId, qualifiedModuleName.Value.Component)
    : (_vbe.ActiveVBProject.ProjectId, _vbe.SelectedVBComponent);
```

File: RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs [hunk 4]
```diff
@@ -154,15 +156,58 @@ private SearchResultsViewModel CreateViewModel(Declaration declaration)
...
+                }
+
+                var selectedCount = component.SelectedControls.Count;                
+                if (selectedCount > 1) { return null; }
+
+                // Cannot use DeclarationType.UserForm, parser only assigns UserForms the ClassModule flag
+                (var selectedType, var selectedName)
+                    = selectedCount == 0
+                                     ? (DeclarationType.ClassModule, component.Name)
+                                     : (DeclarationType.Control, component.SelectedControls[0].Name);
```
Comment (retailcoder) [inline RetailCoder.VBE/UI/Command/FindAllReferencesCommand.cs]: Again - I'd put the assignment on the same line as the tuple definition, and have the ternaries at only 1 level of indentation:

```csharp
(var selectedType, var selectedName) = selectedCount == 0
    ? (DeclarationType.ClassModule, component.Name)
    : (DeclarationType.Control, component.SelectedControls[0].Name);
```

**Pull request merged**